### PR TITLE
Fixes for mixed precision training

### PR DESCRIPTION
**Pull request by epwalsh**

### CODE REVIEW

File: allennlp_models/rc/models/bidaf.py [hunk 1]
```diff
@@ -275,8 +275,17 @@ def forward(  # type: ignore
         span_end_input = self._dropout(torch.cat([final_merged_passage, encoded_span_end], dim=-1))
         span_end_logits = self._span_end_predictor(span_end_input).squeeze(-1)
         span_end_probs = util.masked_softmax(span_end_logits, passage_mask)
-        span_start_logits = util.replace_masked_values(span_start_logits, passage_mask, -1e7)
-        span_end_logits = util.replace_masked_values(span_end_logits, passage_mask, -1e7)
+
+        # Replace the masked values with a very negative constant.
+        # We have to be careful here because if we're training with half/mixed precision floats,
+        # -1e7 would result in an overflow. Hence we fall back to the min FP16 value.
+        very_negative_constant = max(torch.finfo(span_start_logits.dtype).min, -1e7)
```
Comment (matt-gardner) [inline allennlp_models/rc/models/bidaf.py:282]: I'm pretty sure we have a utility function for this. Looking it up...

Comment (matt-gardner) [reply allennlp_models/rc/models/bidaf.py:282]: These: https://github.com/allenai/allennlp/blob/33a49454db239bfe3de6928229c405da056a8bf4/allennlp/nn/util.py#L1859-L1899.

File: tests/rc/models/bidaf_test.py [hunk 2]
```diff
@@ -132,3 +138,15 @@ def test_mismatching_dimensions_throws_configuration_error(self):
...
+
+@requires_gpu
+@pytest.mark.skipif(amp is None, reason="Apex is not installed.")
+class BidirectionalAttentionFlowMixedPrecisionTest(AllenNlpTestCase):
+    @flaky(max_runs=5)
+    def test_model_can_train_save_and_load_with_mixed_precision(self):
+        train_model_from_file(
+            FIXTURES_ROOT / "rc" / "bidaf" / "experiment.json",
+            self.TEST_DIR,
+            overrides="{'trainer.opt_level':'O1','trainer.cuda_device':0}",
```
Comment (matt-gardner) [inline tests/rc/models/bidaf_test.py:151]: Given what you recently said in slack, should this be O2?

File: tests/rc/models/transformer_qa_test.py [hunk 3]
```diff
@@ -49,3 +56,14 @@ def test_forward_pass_runs_correctly(self):
...
+
+
+@requires_gpu
+@pytest.mark.skipif(amp is None, reason="Apex is not installed.")
+class TrasnformerQaMixedPrecisionTest(AllenNlpTestCase):
+    def test_model_can_train_save_and_load_with_mixed_precision(self):
+        train_model_from_file(
+            FIXTURES_ROOT / "rc" / "transformer_qa" / "experiment.jsonnet",
+            self.TEST_DIR,
+            overrides="{'trainer.opt_level':'O1','trainer.cuda_device':0}",
```
Comment (matt-gardner) [inline tests/rc/models/transformer_qa_test.py:68]: Same here about O2.

File: tests/rc/models/transformer_qa_test.py [hunk 4]
```diff
@@ -49,3 +56,14 @@ def test_forward_pass_runs_correctly(self):
         assert span_start <= span_end
         assert span_end < self.instances[0].fields["question_with_context"].sequence_length()
         assert isinstance(output_dict["best_span_str"][0], str)
+
+
+@requires_gpu
+@pytest.mark.skipif(amp is None, reason="Apex is not installed.")
+class TrasnformerQaMixedPrecisionTest(AllenNlpTestCase):
```
Comment (matt-gardner) [inline tests/rc/models/transformer_qa_test.py:63]:
```suggestion
class TransformerQaMixedPrecisionTest(AllenNlpTestCase):
```

File: allennlp_models/rc/models/dialog_qa.py [hunk 5]
```diff
@@ -515,7 +518,9 @@ def _get_best_span_yesno_followup(
         if span_start_logits.dim() != 2 or span_end_logits.dim() != 2:
             raise ValueError("Input shapes must be (batch_size, passage_length)")
         batch_size, passage_length = span_start_logits.size()
-        max_span_log_prob = [-1e20] * batch_size
+
+        very_negative_constant = max(torch.finfo(span_start_logits.dtype).min, -1e20)
```
Comment (matt-gardner) [inline allennlp_models/rc/models/dialog_qa.py:522]: This one can still just call [this function](https://github.com/allenai/allennlp/blob/32bccfbdaf97045f31861ab16bcfdefb8007c3f2/allennlp/nn/util.py#L1871), right?

File: allennlp_models/rc/models/utils.py [hunk 6]
```diff
@@ -30,3 +32,16 @@ def get_best_span(span_start_logits: torch.Tensor, span_end_logits: torch.Tensor
...
+def replace_masked_values_with_big_negative_number(
+    x: torch.Tensor, mask: torch.Tensor, default: float = -1e7
+):
+    """
+    Replace the maskedd values in a tensor something really negative so that they won't
+    affect a max operation.
+    """
+    # We have to be careful here because if we're training with half/mixed precision floats,
+    # -1e7 would result in an overflow. Hence we fall back to the min FP16 value.
+    big_negative_num = max(torch.finfo(x.dtype).min, default)
```
Comment (matt-gardner) [inline allennlp_models/rc/models/utils.py:46]: We already have [a function for this](https://github.com/allenai/allennlp/blob/32bccfbdaf97045f31861ab16bcfdefb8007c3f2/allennlp/nn/util.py#L1871); why not use it?  I think it's fine to add this new outer function to simplify model logic, I just don't understand this line.

Comment (epwalsh) [reply allennlp_models/rc/models/utils.py:46]: I wasn't aware of that function

Comment (matt-gardner) [reply allennlp_models/rc/models/utils.py:46]: No problem; that's what I was pointing to with my first comment way up above.  Sorry it wasn't clear enough.

**Comment by epwalsh** [comment]
I'm going to try training TransformerQA with opt level O2 / O3 again before merging this

**Pull request merged**